snap/channel/channel.go

```diff
@@ -200,3 +211,53 @@ func (c *Channel) Match(c1 *Channel) ChannelMatch {
 		Risk:         requestedRiskLevel >= rl1,
 	}
 }
+
+// Resolve resolves newChannel wrt channel, this means if newChannel
+// is risk/branch only it will preserve the track of channel. It
+// assumes that if both are not empty, channel is parseable.
+func Resolve(channel, newChannel string) (string, error) {
+	if newChannel == "" {
+		return channel, nil
+	}
+	if channel == "" {
+		return newChannel, nil
+	}
+	ch, err := ParseVerbatim(channel, "-")
+	if err != nil {
+		return "", err
+	}
+	p := strings.Split(newChannel, "/")
+	// risk/branch only
+	if strutil.ListContains(channelRisks, p[0]) && ch.Track != "" {
+		return ch.Track + "/" + newChannel, nil
+	}
+	return newChannel, nil
+}
+
+var ErrTrackSwitch = errors.New("cannot switch locked track")
```

Comment:
Nitpick: the error name is a bit generic (any kind of track switch operation can claim this name). I'd suggest using something matching `Locked` name in `ResolveLocked`.